File: frontend/src/lib/chapterIndex.js
export const CHAPTER_INDEX_CONTRACT_VERSION = 'chapter-index.v1';

const STRUCTURED_TITLE = /^(chapter|letter|book|part|section|canto|volume|poem|gitanjali)\s+([ivxlcdm]+|\d+)[.:]?\s*(.*)$/i;
const ACRONYMS = new Set(['ai', 'api', 'css', 'dna', 'html', 'pdf', 'uk', 'us', 'usa']);

function romanToArabic(value = '') {
  const text = String(value || '').toUpperCase();
  if (!/^[IVXLCDM]+$/.test(text)) return null;
  const values = { I: 1, V: 5, X: 10, L: 50, C: 100, D: 500, M: 1000 };
  let total = 0;
  let previous = 0;
  [...text].reverse().forEach((character) => {
    const current = values[character];
    total += current < previous ? -current : current;
    previous = Math.max(previous, current);
  });
  return total || null;
}

function smartTitleCaseSegment(value = '') {
  const text = String(value || '').trim();
  const letters = text.match(/[A-Za-z]/g) || [];
  const upperLetters = text.match(/[A-Z]/g) || [];
  if (!letters.length || upperLetters.length / letters.length < 0.72) return text;
  return text
    .toLowerCase()
    .replace(/\b([a-z])([a-z'’.]*)/g, (word, first, rest) => (
      ACRONYMS.has(word.toLowerCase()) ? word.toUpperCase() : `${first.toUpperCase()}${rest}`
    ))
    .replace(/\b(Dr|Mr|Mrs|Ms)\b\.?/gi, (_match, honorific) => `${honorific[0].toUpperCase()}${honorific.slice(1).toLowerCase()}.`);
}

export function normalizeChapterDisplayTitle(title = '') {
  const original = String(title || '').trim();
  if (!original) return '';
  const cleaned = original
    .replace(/[_*`]+/g, '')
    .replace(/\s*[.:]?\s*(?:(?:--|—|-)\s*)?continued\.?\s*$/i, '')
    .replace(/\s+/g, ' ')
    .trim();
  const match = cleaned.match(STRUCTURED_TITLE);
  if (!match) return smartTitleCaseSegment(cleaned);
  const [, unit, numeral, remainder] = match;
  const number = /^\d+$/.test(numeral) ? Number(numeral) : (romanToArabic(numeral) || numeral.toUpperCase());
  const unitLabel = `${unit[0].toUpperCase()}${unit.slice(1).toLowerCase()} ${number}`;
  const subtitle = smartTitleCaseSegment(remainder);
  return subtitle ? `${unitLabel}. ${subtitle}` : unitLabel;
}

export function chapterIndexEntry(chapter = {}, position = 1, total = 1) {
  const apiContractCurrent = chapter.index_contract === CHAPTER_INDEX_CONTRACT_VERSION;
  if (apiContractCurrent && chapter.index_title && chapter.index_sequence_label) {
    return chapter;
  }
  const displayTitle = normalizeChapterDisplayTitle(chapter.display_title || chapter.title);
  const match = displayTitle.match(STRUCTURED_TITLE);
  const unitLabel = match ? `${match[1][0].toUpperCase()}${match[1].slice(1).toLowerCase()} ${match[2]}` : '';
  const subtitle = match ? match[3].trim().replace(/^[.:—-]\s*/, '') : '';
  const width = Math.max(2, String(Math.max(Number(total) || 1, 1)).length);
  return {
    ...chapter,
    index_contract: CHAPTER_INDEX_CONTRACT_VERSION,
    index_sequence: Math.max(Number(position) || 1, 1),
    index_sequence_label: String(Math.max(Number(position) || 1, 1)).padStart(width, '0'),
    display_title: displayTitle,
    index_title: subtitle || unitLabel || displayTitle || `Section ${position}`,
    index_secondary_label: subtitle ? unitLabel : '',
  };
}

export function sortedChapterIndex(chapters = []) {
  return chapters
    .map((chapter, sourceIndex) => ({ chapter, sourceIndex }))
    .sort((left, right) => (
      (Number(left.chapter?.order) || 0) - (Number(right.chapter?.order) || 0)
      || String(left.chapter?.id || '').localeCompare(String(right.chapter?.id || ''))
      || left.sourceIndex - right.sourceIndex
    ))
    .map(({ chapter }, index, sorted) => chapterIndexEntry(chapter, index + 1, sorted.length));
}

File: backend/domain/chapter_index.py
"""Deterministic chapter-index labels shared by reader API surfaces."""

from __future__ import annotations

import re
from typing import Any, Iterable, Mapping


CHAPTER_INDEX_CONTRACT_VERSION = "chapter-index.v1"

_FORMAT_MARKS_RE = re.compile(r"[_*`]+")
_CONTINUATION_SUFFIX_RE = re.compile(
    r"\s*[.:]?\s*(?:(?:--|—|-)\s*)?continued\.?\s*$",
    re.IGNORECASE,
)
_STRUCTURED_TITLE_RE = re.compile(
    r"^(chapter|letter|book|part|section|canto|volume|poem|gitanjali)\s+"
    r"([ivxlcdm]+|\d+)[.:]?\s*(.*)$",
    re.IGNORECASE,
)
_ALL_CAPS_WORD_RE = re.compile(r"\b([a-z])([a-z'’.]*)", re.IGNORECASE)
_HONORIFIC_RE = re.compile(r"\b(Dr|Mr|Mrs|Ms)\b\.?", re.IGNORECASE)
_ACRONYMS = {"ai", "api", "css", "dna", "html", "pdf", "uk", "us", "usa"}


def _roman_to_int(value: str) -> int | None:
    text = str(value or "").upper()
    if not text or not re.fullmatch(r"[IVXLCDM]+", text):
        return None
    values = {"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}
    total = 0
    previous = 0
    for character in reversed(text):
        current = values[character]
        total += -current if current < previous else current
        previous = max(previous, current)
    return total if total > 0 else None


def _smart_title_case(value: str) -> str:
    text = str(value or "").strip()
    letters = re.findall(r"[A-Za-z]", text)
    uppercase = re.findall(r"[A-Z]", text)
    if not letters or len(uppercase) / len(letters) < 0.72:
        return text

    def replace_word(match: re.Match[str]) -> str:
        word = match.group(0)
        lowered = word.lower()
        if lowered in _ACRONYMS:
            return lowered.upper()
        return f"{lowered[0].upper()}{lowered[1:]}"

    normalized = _ALL_CAPS_WORD_RE.sub(replace_word, text.lower())
    return _HONORIFIC_RE.sub(lambda match: f"{match.group(1).title()}.", normalized)


def normalize_chapter_display_title(title: Any) -> str:
    """Return a reader-safe title without mutating source metadata."""

    original = str(title or "").strip()
    if not original:
        return ""
    cleaned = _FORMAT_MARKS_RE.sub("", original)
    cleaned = _CONTINUATION_SUFFIX_RE.sub("", cleaned)
    cleaned = re.sub(r"\s+", " ", cleaned).strip()
    match = _STRUCTURED_TITLE_RE.match(cleaned)
    if not match:
        return _smart_title_case(cleaned)
    unit, numeral, remainder = match.groups()
    number = int(numeral) if numeral.isdigit() else (_roman_to_int(numeral) or numeral.upper())
    unit_label = f"{unit.title()} {number}"
    subtitle = _smart_title_case(remainder)
    return f"{unit_label}. {subtitle}" if subtitle else unit_label


def chapter_index_entry(
    chapter: Mapping[str, Any],
    *,
    position: int,
    total: int,
) -> dict[str, Any]:
    """Decorate one chapter with a stable visual/accessibility index contract."""

    source_title = str(chapter.get("title") or "").strip()
    display_title = normalize_chapter_display_title(source_title)
    match = _STRUCTURED_TITLE_RE.match(display_title)
    unit_label = ""
    subtitle = ""
    if match:
        unit, numeral, remainder = match.groups()
        unit_label = f"{unit.title()} {numeral}"
        subtitle = remainder.strip().lstrip(".:-— ").strip()

    width = max(2, len(str(max(int(total or 0), 1))))
    sequence_label = str(max(int(position or 0), 1)).zfill(width)
    primary_title = subtitle or unit_label or display_title or f"Section {position}"
    secondary_label = unit_label if subtitle else ""
    return {
        **dict(chapter),
        "index_contract": CHAPTER_INDEX_CONTRACT_VERSION,
        "index_sequence": max(int(position or 0), 1),
        "index_sequence_label": sequence_label,
        "display_title": display_title or primary_title,
        "index_title": primary_title,
        "index_secondary_label": secondary_label,
    }


def build_chapter_index_entries(chapters: Iterable[Mapping[str, Any]]) -> list[dict[str, Any]]:
    """Sort and decorate a complete chapter list deterministically."""

    indexed = list(enumerate(chapters))
    indexed.sort(
        key=lambda pair: (
            int(pair[1].get("order", pair[0]) or 0),
            str(pair[1].get("id") or ""),
            pair[0],
        )
    )
    total = len(indexed)
    return [
        chapter_index_entry(chapter, position=position, total=total)
        for position, (_source_index, chapter) in enumerate(indexed, start=1)
    ]
